Enable the AsnXml build target on non-Windows.

vcsjones opened the pull request:
This enables the `AsnXml` target to run on all platforms.

vcsjones committed:
Enable the AsnXml build target on non-Windows.

vcsjones commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AsnXml.targets:
```diff
@@ -8,10 +8,8 @@
-  <!-- MSBuild for .NET Core doesn't support XslTransform at this time -->
   <Target Name="CompileAsn" BeforeTargets="CoreCompile"
-    Condition="'$(MSBuildRuntimeType)' != 'core'"
-    Inputs="@(AsnXml);$(MSBuildThisFileDirectory)asn.xslt"
```
Removing `;$(MSBuildThisFileDirectory)asn.xslt` seemed like the right thing to do? With it there, the `Target` would run even if `@(AsnXml)` was empty, and the `XslTransformation` task would fail saying it wasn't given any input.

bartonjs replied:
Well, if the xslt isn't listed as an input then changing the xslt doesn't invalidate the target.  The fix is (I think) the above suggestion.

vcsjones replied:
Ah, okay. Got it.

vcsjones commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AsnXml.targets:
```diff
@@ -21,23 +19,25 @@
-    
-    <Warning Condition="'@(_AsnXmlDiffCode)' != '0'" Text="AsnXml regenerated files, be sure to check them in: @(AsnXml -> '%(Identity).cs')" />
+
+    <Message Condition="'@(_AsnXmlDiffCode)' != '0'" Text="AsnXml regenerated files, be sure to check them in: @(AsnXml -> '%(Identity).cs')" Importance="high" />
```
Since this now runs as part of the normal build (`./build.sh` from the root) and we treat warnings as errors there, the warning would fail the build if the files were regenerated. The message stops that from happening, but unfortunately is not quite as prominent and easy to miss.

bartonjs replied:
Warning (as error) might be better, since it means that the source tree and the build are not in a consistent state.  If you've changed one of these files it means you have to up-enter.  If you forget to add one to a commit you fail the build in the PR.

vcsjones replied:
> Warning (as error) might be better

Hm, it seems strange to fail the build just to remind you to check in a file. Seems like friction for someone less familiar with how it works.

> If you've changed one of these files it means you have to up-enter.

I don't follow. You need to run the build twice if you change one of these files?

> If you forget to add one to a commit you fail the build in the PR.

Well, that's true of normal code changes as well assuming a test is written to cover the changes.

vcsjones replied:
Hmm. I thought on this a bit more and it makes some sense now if you think about it the way the reference source projects work. Ideally if the XML and .cs file are out-of-sync you want to fail in CI (which is what I think you were saying) much like forgetting to update / check in the ref source.

vcsjones replied:
In that case I would want to go full `<Error />` instead of `<Warning />` so that the error consistently happens whether you use `./build.sh` or `dotnet build` in a project directory (the latter is what I tend to use).

bartonjs replied:
Ah, but that's the beauty of Warning with Warn-As-Error.  When you're developing (and use `dotnet build` locally) it's not an error, just a "hey, pay attention" blob of yellow.  If you don't listen and try to check in then it's an error.

So I think Warning (with Warn-As-Error from CI) is still the good place to be.

If there are complaints we can adjust it to message normally and Error when some variable tells us it's an official build.  (Then you could still use `build.{cmd|sh} -allConfigurations -subset libs` for "I changed the xslt, regenerate everything" but `build.{cmd|sh} -ci` would produce an error in that situation.)  What I didn't like about that is that switching on the officialness of the build is a little fragile (if the variable changes and this file is missed, it just never sees an official build).  So maybe the increment would be it's a message for debug builds and a warning or error for release builds.

vcsjones replied:
Okay, I went back to a warning for now.

vcsjones commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AsnXml.targets:
```diff
@@ -21,23 +19,25 @@
+    <Exec Condition="$([MSBuild]::IsOsPlatform('windows')) != 'true'"
+      IgnoreExitCode="true"
+      StandardOutputImportance="Low"
+      Command="/usr/bin/diff -q -a @(AsnXml -> '$(IntermediateOutputPath)asnxml\%(filename).cs') @(AsnXml -> '%(Identity).cs')">
```
I thiiiiink `/usr/bin/diff` on all non-Windows platforms? It is on Ubuntu Linux and macOS. It felt a little better to me to specify a full path instead.

am11 replied:
I think it would be portable if we add hardcoded path as a fallback to `command(1)`:

```suggestion
      Command="%24%28command -v diff || echo /usr/bin/diff%29 -q -a @(AsnXml -> '$(IntermediateOutputPath)asnxml\%(filename).cs') @(AsnXml -> '%(Identity).cs')">
```

vcsjones replied:
Hmm, isn't `command` a dash/bash builtin? Are there any concerns around this not being portable?

am11 replied:
`command -v diff` is POSIX shelll way of finding the location. https://pubs.opengroup.org/onlinepubs/009604499/utilities/command.html
Depending on package manager / distro / OS, we can have `diff(1)` installed in `/usr/local/bin` or under `/opt` etc.

vcsjones replied:
Gotcha. Thank you!

vcsjones reviewed (commented):
I'm in "I have no idea what I'm doing" territory with MSBuild, so if something seems weird, that's why.

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AsnXml.targets:
```diff
@@ -8,10 +8,8 @@
 
-  <!-- MSBuild for .NET Core doesn't support XslTransform at this time -->
   <Target Name="CompileAsn" BeforeTargets="CoreCompile"
-    Condition="'$(MSBuildRuntimeType)' != 'core'"
```
Somewhere along the way I ended up with the target file firing when it made no sense to do so.  It might be worth keeping a condition, but instead have

```
Condition=" '@(AsnXml)' != '' "
```

vcsjones replied:
Seeing as we both commented at the same time and you may have written this before I posted https://github.com/dotnet/runtime/pull/45220#discussion_r530572836, is my change there sufficient or is is the condition still desirable?

vcsjones committed:
Fix condition and input

am11 commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AsnXml.targets:
```diff
@@ -21,23 +20,25 @@
       OutputPaths="@(AsnXml -> '$(IntermediateOutputPath)asnxml\%(filename).cs')" />
 
-    <Exec Condition="'$(TargetOS)'=='windows'"
+    <Exec Condition="$([MSBuild]::IsOsPlatform('windows')) == 'true'"
```
a nit: in other places, the prevalent pattern is: `Condition="'$(TargetsWindows)' == 'true'"` (and TargetsUnix below). Could that work here?

vcsjones replied:
To help me better understand correctly: I thought that `TargetOS` and `TargetsWindows` are about what target you are building for, not the actual OS that MSBuild is running on. Regardless of the OS we are targeting to build; we want to use the `fc` tool if we are building on WIndows.

bartonjs replied:
Your understanding matches mine.  Probably means that the "hey, I updated something" warnings just didn't fire when building Unix-y projects from Windows.

am11 replied:
Ah, correct. I didn't realized that we have `build on Windows, targeting Unix` support. :)

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AsnXml.targets:
```diff
@@ -21,23 +20,25 @@
-    <ItemGroup Condition="'$(TargetOS)'!='windows'">
-      <_AsnXmlDiffCode Include="1" />
-    </ItemGroup>
+    <Exec Condition="$([MSBuild]::IsOsPlatform('windows')) != 'true'"
```
If this is the same as the Windows one except for the command, maybe

```msbuild
    <PropertyGroup>
      <_AsnXmlDiffCmd>/usr/bin/diff -q -a</_AsnXmlDiffCmd>
      <_AsnXmlDiffCmd Condition="$([MSBuild]::IsOsPlatform('windows') == 'true'">$(SystemRoot)\System32\fc.exe /a</_AsnXmlDiffCmd>
    </PropertyGroup>

    <Exec
      IgnoreExitCode="true"
      StandardOutputImportance="Low"
      Command="$(_AsnXmlDiffCmd) @(AsnXml -> '$(IntermediateOutputPath)asnxml\%(filename).cs') @(AsnXml -> '%(Identity).cs')">
      <Output TaskParameter="ExitCode" ItemName="_AsnXmlDiffCode" />
    </Exec>
```

vcsjones committed:
Simply exec based on feedback.

vcsjones committed:
Fix intermediate path.

vcsjones committed:
Use command to find the right diff tool.

Co-authored-by: Adeel Mujahid <adeelbm@outlook.com>

vcsjones commented on src/libraries/Common/src/System/Security/Cryptography/Asn1/AsnXml.targets:
```diff
@@ -1,43 +1,40 @@
     <AsnXml />
+    <_AsnXmlDiffCmd>%24%28command -v diff%29 -q -a</_AsnXmlDiffCmd>
+    <_AsnXmlDiffCmd Condition="$([MSBuild]::IsOsPlatform('windows')) == 'true'">$(SystemRoot)\System32\fc.exe /a</_AsnXmlDiffCmd>
+    <_AsnIntermediatePath>$([MSBuild]::NormalizeDirectory('$(IntermediateOutputPath)', 'asnxml'))</_AsnIntermediatePath>
```
`diff` was getting cranky being passed Windows directory separator when using `$(IntermediateOutputPath)asnxml\%(filename).cs`. Instead I made a variable here which will ensure there is a trailing slash that is appropriate for the current OS.

vcsjones committed:
Back to using Warning behavior

vcsjones commented:
Ha, and now it is failing in CI. Seems that the diffs it is seeing are whitespace or BOM related.

bartonjs commented:
I was kinda wondering about that :smile:.  I expect the easiest answer is to comment out the warning and do `build.whatever -allConfigurations -subset libs -rc Release` (or it might make you do libs+clr), then put the warning back.  Then you'll have all the files fixed locally and can just add them as a catch-up commit.

vcsjones commented:
It's the BOM, but it's doing the opposite of what I expected. The `XslTransformation` is generating with the BOM, but the checked-in files do not have a BOM.

@bartonjs UTF8 files with a BOM seems a little weird to me, but, seems like we can check them in. I don't really see anything pro or against having BOMs in the repo as far as a guideline, ahd the `.editorconfig` file has no opinion on this in C# files. Alternatively we can post-process the generated file and strip the BOM, if we want.

vcsjones committed:
Use UTF8 file encoding with BOM for generated ASN.

Also update the editorconfig to indicate these files are automatically
generated and should have the BOM so that opening them with editors does
not strip out the BOM.

vcsjones committed:
Update PKCS files.

Updates the BOM and functional changes.

vcsjones commented:
@bartonjs I added two more commits. The first is one that only updates the C# files to include the BOM, as the XSLT task does. I also updated the editorconfig to indicate that these files explicitly want the BOM so that if someone opens them in a code editor, some tooling won't try to normalize it.

The second commit automatically has functional changes. It seems that at some point the XSLT file changed but the Pkcs files were not re-generated.

vcsjones commented:
This is fun I'm learning a lot.

Anyway, a question. Editing the XSLT doesn't necessarily re-generate all of the C#. Shouldn't all of the project items have a dependency on the XSLT? e.g.

```xml
    <Compile Include="$(CommonPath)System\Security\Cryptography\Asn1\X509ExtensionAsn.xml.cs">
      <Link>Common\System\Security\Cryptography\Asn1\X509ExtensionAsn.xml.cs</Link>
      <DependentUpon>Common\System\Security\Cryptography\Asn1\X509ExtensionAsn.xml</DependentUpon>
    </Compile>
```

Should be:

```xml
    <Compile Include="$(CommonPath)System\Security\Cryptography\Asn1\X509ExtensionAsn.xml.cs">
      <Link>Common\System\Security\Cryptography\Asn1\X509ExtensionAsn.xml.cs</Link>
      <DependentUpon>Common\System\Security\Cryptography\Asn1\X509ExtensionAsn.xml</DependentUpon>
      <DependentUpon>Common\System\Security\Cryptography\Asn1\asn.xslt</DependentUpon>
    </Compile>
```

? 

When I change the XSLT, I see the target fire, but `AsnXml` property is empty, so we end up not firing the Target in any meaningful way because the `AsnXml` items are up-to-date as far as MSBuild knows without that dependency.

vcsjones committed:
More BOM updates. Remove changes to editorconfig.

The editorconfig support didn't cover everything.

bartonjs commented:
> When I change the XSLT, I see the target fire, but AsnXml property is empty, so we end up not firing the Target in any meaningful way because the AsnXml items are up-to-date as far as MSBuild knows without that dependency.

Strange.  When I originally wrote things it worked... maybe it's a difference in MSBuild and MSBuild.Core?  Certainly there's an easy short-term workaround of `find . -name \*.xml.cs -exec rm {} \;` and then rerun the build; and then looking for a better solution / figuring out why AsnXml is empty can be done a little later (we don't change the xslt all that often).

The generated_code in editorconfig might still be valueable, I think it tells things like stylecop to not run (it's hard to keep all errant spaces out of the xslt-generated output).

bartonjs commented:
Nevermind about the editorconfig; I see that was a thing you added and removed... apparently I had jumped only to the "removed" commit :)

bartonjs commented:
As far as the BOM, I'm happier with it not being present.  So, if you can figure out how to get the xslt thing to not emit a BOM that would make me happier; but there are times when it's just not worth fighting tools.

vcsjones committed:
Missed one, somehow

vcsjones commented:
> So, if you can figure out how to get the xslt thing to not emit a BOM

I don't immediately see one. XSLT 2.0 has [`byte-order-mark`](https://www.w3.org/TR/xslt20/#serialization) on `xsl:output` specifically for this... but we don't support XSLT 2.0. And I don't see a way to feed in custom things to the `XmlWriter`. We could post process with an inline C# task, but I don't see any usages of that in libraries and wonder if it is perhaps not advised.

vcsjones commented:
Hm. I think an issue with `-allConfigurations` and `dotnet build` when you don't specify a specific target framework is that the same XML gets built concurrently.

The effective intermediate path is `artifacts/obj/System.Security.Cryptography.X509Certificates/Debug/asnxml/` for S.S.C.X509Certificates. If allConfigurations builds Unix and Windows at the same time, then the XSL Transform task will try to run concurrently, and one will fail because the file is locked.

We could add the `$(TargetFramework)` as part of the intermediate path and that would fix the XSLT task, but that moves the concurrence problem to the `Copy` task.

bartonjs commented:
> We could add the $(TargetFramework) as part of the intermediate path and that would fix the XSLT task,

Probably also needs the target OS (when split) so it doesn't fail when building net5.0-Linux and net5.0-OSX at the same time.  Though it looks like that should already be taken into account for the IntermediateOutputPath variable... so I'm surprised that there are two builds running concurrently for the same thing.

e.g. I see this when building the Cng project:

Property reassignment: $(IntermediateOutputPath)="C:\git\bartonjs\runtime\artifacts\obj\System.Security.Cryptography.Cng\net5.0-Windows_NT-Debug\" (previous value: "C:\git\bartonjs\runtime\artifacts\obj\System.Security.Cryptography.Cng\Debug\") at C:\git\bartonjs\runtime\eng\BeforeTargetFrameworkInference.targets (12,5)

oooh.... I wonder if the problem goes away if you move that PropertyGroup to inside the Task.  I think this property set evaluates too early and is getting the pre-substitution IntermediateOutputPath (a consequence of factoring it out).

> but that moves the concurrence problem to the Copy task.

Thankfully the Copy task is conditional on the intermediate file having a diff from the target file, so that race wouldn't happen in any build that wasn't changing an input... and since Copy retries (and the two sources are identicial) it just means a local build would (worst case) get one of the "File in use, retrying" warnings if that race hit.

vcsjones committed:
Fix property group

vcsjones committed:
Fix file that needed to be re-generated.

vcsjones commented:
> oooh.... I wonder if the problem goes away if you move that PropertyGroup to inside the Task. I think this property set evaluates too early and is getting the pre-substitution IntermediateOutputPath (a consequence of factoring it out).

I would not have figured that out in a million years.

bartonjs commented:
> I would not have figured that out in a million years.

I've, bizarrely, had to learn a bit about MSBuild in the last few days.  Things like this :smile:.

bartonjs commented:
The Android failures look to be something about the device needing to reboot, maybe?  But they don't look related, so I'll be confident and hit merge.
